### ISSUE TITLE
capnp: zero out memory of cleared pointer fields

### ISSUE THREAD

**Issue by zombiezen** [zombiezen/go-capnproto2#84]
@kentonv mentioned this to me at the 0.6 release party, and addresses an issue @abraithwaite raised to me a while back.  In read-modify-write scenarios where you are changing a pointer field, it is desirable to recursively zero out the objects formerly pointed to.  @kentonv says that it's not necessary to have the region reclaimed by the arena.  This is what the C++ implementation does.

I have two small concerns about this feature.  First, I don't see a good way of allowing a struct to be moved to a different pointer field without introducing more API (at the very least to make an explicit copy).  I don't think this is a common use case, so I don't see this as a huge deal.  Second, this is a behavior change in existing API that can only be caught at runtime.  I'm not sure whether to fix this in v2 or v3.

**Comment by kentonv** [comment]
FWIW in C++ the way you move pointers is through the "orphan" API.

    auto orphan = someStruct.disownSomeField();
    someOtherStruct.adoptSomeField(kj::mv(orphan));

If an orphan goes out-of-scope without being adopted, then that's when the zeroing happens (the destructor calls a method which I named... uh... `euthanize()`...). So, if you just do:

    someStruct.disownSomeField();

ignoring the return value, then the zeroing kicks in.

I guess in Go you don't have RAII or move semantics so you can't really copy this API directly.

**Comment by zombiezen** [comment]
Correct.  The Go API basically acts like the Orphan API in C++ at all times.  There could be a new "clear field shallowly" method and that would allow moves.  How about this set of changes:

1. A new `DisownPtr` method on `Struct`.  It would zero out the pointer field and return the old pointer.
2. A new generated struct method for each pointer field that calls `DisownPtr`.
3. Change the existing `SetPtr` method to do recursive clearing if a non-null value is present (breaking behavior change, but I'm unclear on how much code would be impacted).

**Comment by kentonv** [comment]
Would SetPtr accept null as an argument, in order to recursively zero a field without setting a new value?

**Comment by zombiezen** [comment]
Yes.

**Comment by kentonv** [comment]
Sounds reasonable.

In C++, calling disownFoo() and then ignoring the return value would end up deep-zeroing it. In Go I assume you don't have a way to detect if the return value was unused, so people will need to make some explicit call to deep-zero a pointer after it is returned.

**Comment by zombiezen** [comment]
@kentonv, how does the C++ implementation deal with clearing pointers when capabilities are involved?  Specifically, do you remove the capability from the capability table?  Seems like you can't know whether a capability index is used elsewhere without walking the whole message.

**Comment by kentonv** [comment]
@zombiezen 

* For readers, the message is immutable, so there's no need to support dropping individual pointers.
* For builders (which always start from scratch), the C++ implementation never uses the same index twice. This means that if a message contains two copies of the same capability (which is unusual), the cap table may have duplicate entries, but that's not a big deal.
* When copying from a reader into a builder, the table indexes are re-assigned as each pointer is copied.

**Comment by zombiezen** [comment]
Acknowledged, but the crucial question in there is still: do you remove the capability from the capability table?

**Comment by kentonv** [comment]
Yes.

**Comment by kentonv** [comment]
Just to be clear: It would be a security problem is removing a capability from a message before sending did not actually prevent the receiver from receiving that capability.

**Comment by zombiezen** [comment]
Okay, yes, I understand now.  I'm not sure how or where to document this, but I think we need to document that Cap'n Proto read-modify-write scenarios *must* have a copy step for security purposes.  IIUC this is true for both C++ and Go.

If that's the case, how much do we care about the recursive zeroing?

**Comment by kentonv** [comment]
In C++, it is not possible to read-modify-write without a copy step because you can only read into a Reader, which is immutable.

Well, OK, that's not quite true: there is an obscure API by which you can create a Builder backed by already-initialized memory. But the API is documented as being unsafe unless you trust the input, and there's no particular support for using it with messages that have capabilities (although you could build such support, but then it's up to you to get it right).

**Comment by KubaO** [comment]
Would a compacting garbage compaction step be a useful addition to the runtime? I would open another issue for that then. This would be reasonably efficient due to the "no two pointers pointing to the same object" limitation - what I just yesterday considered a limitation seems to be a very wise decision indeed.

**Comment by kentonv** [comment]
@KubaO A "compacting garbage collection" can already be achieved in O(n) time by making a copy of the message. Only reachable objects will be copied.
